### PR TITLE
VerifyProofs

### PR DESCRIPTION
**Pull request by rianhughes**
The PR implements VerifyRangeProof. It accounts for the cases where either 0, 1 or both boundary proofs are supplied. It does require the proof keys to be set.

Note: It does not account for the case where the proof keys are not set. I have opened a new PR that accounts for this case https://github.com/NethermindEth/juno/pull/1901

### CODE REVIEW

File: core/trie/proof.go [hunk 1]
```diff
@@ -145,6 +146,22 @@ func GetProof(leaf *felt.Felt, tri *Trie) ([]ProofNode, error) {
 	return proofNodes, nil
 }
 
+func GetProofs(startKey, endKey *felt.Felt, tri *Trie) ([][]ProofNode, error) {
+	oneFelt := new(felt.Felt).SetUint64(1)
+	iterKey := startKey
+	leafRange := new(felt.Felt).Sub(endKey, startKey).Uint64()
+	proofs := make([][]ProofNode, leafRange)
+	for i := range leafRange {
```
Comment (asdacap) [inline core/trie/proof.go:154]: How does it iterate over uint64?

File: core/trie/proof.go [hunk 2]
```diff
@@ -175,5 +192,129 @@ func VerifyProof(root *felt.Felt, key *Key, value *felt.Felt, proofs []ProofNode
...
+
+	}
+	return nil, errors.New("failed to get proof value")
+}
+
+// VerifyRangeProof verifies the range proof for the given range of keys.
+// ref: https://github.com/ethereum/go-ethereum/blob/v1.14.3/trie/proof.go#L484
+// Note: this currently assumes that the inner keys do not contain the min/max key (ie both proofs exist) // Todo
+// The first/last key and value must correspond to the left/right proofs //Todo we currently assume both proofs are provided, as above
+func VerifyRangeProof(root *felt.Felt, keys []*Key, values []*felt.Felt, proofs [2][]ProofNode, hash hashFunc) (bool, error) {
```
Comment (asdacap) [inline core/trie/proof.go:243]: Note, the `proofs` is a single array, not two. And the `keys` are `*felt.Felt`, meaning its always leaf. Well, I guess if you add a note on that, it does not matter if its a `Key` or `Felt`.

File: core/trie/proof.go (hunk 1 above)
Comment (rianhughes) [reply core/trie/proof.go:154]:
```
someuint := uint(10)
for i := range someuint {
    fmt.Println(i)
}
```
prints 0 to 9

File: core/trie/proof.go (hunk 2 above)
Comment (rianhughes) [reply core/trie/proof.go]: We only require two proofs no? Where each proof is a set of nodes from the root to a key.
Eg [geth](https://github.com/ethereum/go-ethereum/blob/master/trie/proof.go#L539) requires two proofs in their implementation. This is also how I understood it from watching Daniels snapsync video

**Comment by rianhughes** [comment]
Note: I still need to account for edge cases (eg either single or no proof), but it would be useful to have a set of eyes on this now

File: core/trie/proof.go [hunk 3]
```diff
@@ -172,8 +194,272 @@ func VerifyProof(root *felt.Felt, key *Key, value *felt.Felt, proofs []ProofNode
...
 	return expectedHash.Equal(value)
 }
+
+// VerifyRangeProof verifies the range proof for the given range of keys.
+// This is achieved by constructing a trie from the boundary proofs, and the supplied key-values.
+// If the root of the reconstructed trie matches the supplied root, then the verification passes.
+// If the trie is constructed incorrectly then the root will have an incorrect key(len,path), and value,
+// and therefore it's hash will be incorrect.
+// ref: https://github.com/ethereum/go-ethereum/blob/v1.14.3/trie/proof.go#L484
+func VerifyRangeProof(root *felt.Felt, keys, values []*felt.Felt, proofKeys, proofValues [2]*felt.Felt,
```
Comment (asdacap) [inline core/trie/proof.go:210]: Note, proof is 1 array, not 2. It got deduplicated into one.

Comment (rianhughes) [reply core/trie/proof.go:210]: A single proof is an array, but we need two proofs, right?

Comment (asdacap) [reply core/trie/proof.go:210]: Well, the easiest way to generate the proof is with the left (start query actually, probably will come back later) and right key separately. Then it is combined into one set. I'm guessing this is done because a lot of the time some of the top nodes are going to be duplicated.

File: core/trie/proof.go [hunk 4]
```diff
@@ -172,8 +194,272 @@ func VerifyProof(root *felt.Felt, key *Key, value *felt.Felt, proofs []ProofNode
...
+		}
+		offset += squishedParent
+
+		// Set the value of the current node
+		crntNode.Value = proofNodes[i].Hash(hashF)
+
+		// Set the children of the current node
+		// The child may need compressed. If so, point to its compressed form.
+		childIdx := i + squishedParent + 1
+		if i+2+squishedParent < len(proofNodes)-1 {
```
Comment (pnowosie) [inline core/trie/proof.go:347]: This if-else here is pure magic - I cannot understand why we're looking at 3 different cases:
- `i+2+squishedParent < len(proofNodes)-1`
- `i+1+offset == len(proofNodes)-1`
- otherwise

also I feel me might want to refactor the `crntNode` construction into a functions which would do the `leafKey.Test(...)` because a lot of code is duplicated or similar and it's hard to follow what happens

**Review by pnowosie — CHANGES_REQUESTED** [review]

File: core/trie/proof.go [hunk 5]
```diff
@@ -87,7 +143,7 @@ func transformNode(tri *Trie, parentKey *Key, sNode storageNode) (*Edge, *Binary
 	}
 
 	rightHash := rNode.Value
-	if isEdge(sNode.key, storageNode{node: rNode, key: sNode.node.Right}) {
+	if isEdge(sNode.key, StorageNode{node: rNode, key: sNode.node.Right}) {
```
[the remark this answers is not in the document]
Comment (rianhughes) [reply core/trie/proof.go:145]: 2. Yes, the value is irrelevant here since we just want to know if the left/right child is an edge node, which we can tell using the length of the parent and child keys

File: core/trie/proof.go (hunk 4 above)
Comment (rianhughes) [reply core/trie/proof.go]: I added some comments and refactored a bit. Let me know if it's still unclear!

File: core/trie/proof.go (hunk 3 above)
Comment (rianhughes) [reply core/trie/proof.go:209]: Geth seems to store proofs nodes in a key-value store, so I guess starkware are thinking about dong something similar.

Comment (rianhughes) [reply core/trie/proof.go:209]: Mind if I open a separate issue for this?

Comment (rianhughes) [reply core/trie/proof.go:209]: https://github.com/NethermindEth/juno/issues/1898

File: core/trie/proof.go (hunk 4 above)
Comment (rianhughes) [reply core/trie/proof.go]: Just an update: this logic is / will be simplified in [this](https://github.com/NethermindEth/juno/pull/1901/files) PR

**Comment by rianhughes** [comment]
This PR is / can be replaced by this https://github.com/NethermindEth/juno/pull/1901

File: core/trie/proof.go (hunk 4 above)
Comment (pnowosie) [reply core/trie/proof.go]: No need to push it here, we can review the changes in PR #1901

**Comment by rianhughes** [comment]
closing for https://github.com/NethermindEth/juno/pull/1901
